### PR TITLE
SALTO-1890: Managed package change validator blocks valid changes

### PR DESCRIPTION
**Pull request by tamtamirr**
The change validator of Salesforce installed packages, used to block all the changes coming from them, even when the changes are valid. (Like editing description or help-text)

---

Added a list that specifies the name of the allowed modifiable properties.

---
_Release Notes_: 
Salesforce Adapter:
- Allow modification of `description` or `help-text` on fields and objects of installed packages.

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/salesforce-adapter/src/change_validators/package.ts [hunk 1]
```diff
@@ -60,6 +74,17 @@ const packageChangeError = async (
   })
 }
 
+const MODIFIABLE_PROPERTIES = [
+  'inlineHelpText',
+  'description',
+]
+
+const isInvalidModify = (change: Change<ChangeDataType>): boolean => {
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/change_validators/package.ts:82]: Is there a better way to do that? @ori-moisis

File: packages/salesforce-adapter/test/change_validators/package.test.ts [hunk 2]
```diff
@@ -158,19 +158,35 @@ describe('package change validator', () => {
...
-        afterField.annotations.modifyMe = 'modified'
-        afterField.refType = createRefToElmWithValue(Types.primitiveDataTypes.MasterDetail)
-        const changeErrors = await packageValidator(
-          [toChange({ before: beforeField, after: afterField })]
-        )
-        expect(changeErrors).toHaveLength(1)
-        expect(changeErrors[0].severity).toEqual('Error')
-        expect(changeErrors[0].elemID).toEqual(beforeField.elemID)
+    describe('modify field with namespace', () => {
+      describe('when modifying a forbidden property', () => {
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/change_validators/package.test.ts:162]: This was the previous test, just added another test scenario for modification of `description`

File: packages/salesforce-adapter/test/change_validators/package.test.ts [hunk 3]
```diff
@@ -158,19 +158,35 @@ describe('package change validator', () => {
...
+        it('should not have change error', async () => {
+          obj.annotate({ [API_NAME]: 'ObjectName__c' })
+          const beforeField = createField(obj, Types.primitiveDataTypes.Lookup, `${obj.annotations[API_NAME]}.MyNamespace__FieldName__c`, undefined, 'description')
+          const afterField = beforeField.clone()
+          afterField.annotations.modifyMe = 'modified'
+          afterField.refType = createRefToElmWithValue(Types.primitiveDataTypes.MasterDetail)
+          const changeErrors = await packageValidator(
+            [toChange({ before: beforeField, after: afterField })]
+          )
+          expect(changeErrors).toBeEmpty()
```
Comment (ori-moisis) [inline packages/salesforce-adapter/test/change_validators/package.test.ts:188]: if this test passes, it means the code is incorrect
this is allowing a change of a field type and an annotation in a field that would not work if we attempted to deploy it to a managed package field

File: packages/salesforce-adapter/src/change_validators/package.ts [hunk 4]
```diff
@@ -60,6 +74,17 @@ const packageChangeError = async (
...
+  const [before, after] = getAllChangeData(change)
+  const modifiedFieldNames = detailedCompare(before, after).map(valueDiff => valueDiff.id.name)
+  return !modifiedFieldNames.every(fieldName => MODIFIABLE_PROPERTIES.includes(fieldName))
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/change_validators/package.ts:85]: I get the feeling that the logic here is incorrect though given the test that passes when it should not

File: packages/salesforce-adapter/src/change_validators/package.ts [hunk 5]
```diff
@@ -13,9 +13,23 @@
...
+  InstanceElement,
+  isModificationChange,
+  isRemovalChange,
+  ChangeError,
+  ChangeValidator,
+  ActionName,
+  isInstanceChange,
+  isFieldChange,
+  isObjectType,
+  getAllChangeData, Change, isAdditionChange, ChangeDataType,
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/package.ts:28]: This is how my IDE adds imports, will change

File: packages/salesforce-adapter/src/change_validators/package.ts [hunk 6]
```diff
@@ -71,9 +96,22 @@ const changeValidator: ChangeValidator = async changes => {
...
 
+  const modifyErrors = await awu(changes)
+    .filter(async change => await isCustomObject(getChangeData(change)) || isFieldChange(change))
+    .filter(change => hasNamespace(getChangeData(change)))
+    .filter(change => isModificationChange(change))
+    .filter(isInvalidModify)
+    .map(change => packageChangeError(
+      change.action,
+      getChangeData(change),
+      `Modification is forbidden of any property that is not one of the following: ${MODIFIABLE_PROPERTIES.join(', ')}`,
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/package.ts:111]: Actually, our message is more informative IMO. This is the error from Salesforce:

Screen Shot 2022-08-16 at 13 47 23

[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/package.ts:111]: Actually, I think the error message from Salesforce is too vague, and unclear. This is an example:

[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/salesforce-adapter/src/change_validators/package.ts:111]: I would keep the prefix at least, something like
```
Cannot modify managed object <object name>.<annotation name>. modification only possible for <valid things>
```
this is important for people who may be searching for answers on google, the salesforce error may have some relevant resources and we'd like our users to find.
by showing a completely different error message, we are preventing our users from finding salesforce answers

File: packages/salesforce-adapter/src/change_validators/package.ts [hunk 7]
```diff
@@ -60,6 +74,17 @@ const packageChangeError = async (
...
 }
 
+const MODIFIABLE_PROPERTIES = [
+  'inlineHelpText',
+  'description',
+]
+
+const isInvalidModify = (change: Change<ChangeDataType>): boolean => {
+  const [before, after] = getAllChangeData(change)
+  const modifiedFieldNames = detailedCompare(before, after).map(valueDiff => valueDiff.id.name)
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/package.ts]: v

File: packages/salesforce-adapter/src/change_validators/package.ts [hunk 8]
```diff
@@ -71,9 +96,22 @@ const changeValidator: ChangeValidator = async changes => {
   const addRemoveErrors = await awu(changes)
     .filter(async change => await isCustomObject(getChangeData(change)) || isFieldChange(change))
     .filter(change => hasNamespace(getChangeData(change)))
+    .filter(change => isAdditionChange(change) || isRemovalChange(change))
     .map(change => packageChangeError(change.action, getChangeData(change)))
     .toArray()
 
+  const modifyErrors = await awu(changes)
+    .filter(async change => await isCustomObject(getChangeData(change)) || isFieldChange(change))
+    .filter(change => hasNamespace(getChangeData(change)))
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/package.ts]: v

File: packages/salesforce-adapter/src/change_validators/package.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/package.ts]: v

File: packages/salesforce-adapter/test/change_validators/package.test.ts (hunk 3 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/test/change_validators/package.test.ts]: v

File: packages/salesforce-adapter/src/change_validators/package.ts [hunk 9]
```diff
@@ -68,12 +119,39 @@ const isInstalledPackageVersionChange = async (
...
+  const installedPackagesChanges = await awu(changes)
     .filter(change => hasNamespace(getChangeData(change)))
+    .toArray()
+  const modificationInstalledPackagesChanges = await awu(installedPackagesChanges)
+    .filter(isModificationChange)
+    .toArray()
+
+  const addRemoveErrors = await awu(installedPackagesChanges)
+    .filter(async change => await isCustomObject(getChangeData(change)) || isFieldChange(change))
+    .filter(change => isAdditionChange(change) || isRemovalChange(change))
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/change_validators/package.ts:131]:
```suggestion
    .filter(isAdditionOrRemovalChange)
```

**Review by ori-moisis — APPROVED** [review]
LGTM

**Pull request merged**